keras_retinanet/losses.py

```diff
@@ -20,48 +20,62 @@
-        classification = keras.backend.reshape(y_pred, (-1, keras.backend.shape(y_pred)[2]))
 
-        # filter out "ignore" anchors
-        anchor_state   = keras.backend.max(labels, axis=1)  # -1 for ignore, 0 for background, 1 for object
-        indices        = keras_retinanet.backend.where(keras.backend.not_equal(anchor_state, -1))
-        classification = keras_retinanet.backend.gather_nd(classification, indices)
-        labels         = keras_retinanet.backend.gather_nd(labels, indices)
-        anchor_state   = keras_retinanet.backend.gather_nd(anchor_state, indices)
+        # compute the divisor: for each image in the batch, we want the number of positive anchors
+
+        # override the -1 labels, since we treat -1/0 the same way for determining the divisor
+        divisor   = keras_retinanet.backend.where(y_true < 0, keras.backend.zeros_like(y_true), y_true)
+        divisor   = keras.backend.max(divisor, axis=2)
+        divisor   = keras.backend.cast(divisor, keras.backend.floatx())
+        # compute the number of positive anchors
+        divisor   = keras.backend.sum(divisor, 1)
+        #  ensure we do not divide by 0
+        divisor   = keras.backend.maximum(1.0, divisor)
+        # pad the tensor to have shape (batch_size, 1, 1) for future division
+        divisor   = keras.backend.expand_dims(divisor, axis=1)
+        divisor   = keras.backend.expand_dims(divisor, axis=2)
+
+        labels         = y_true
+        classification = y_pred
 
-        # select classification scores for labeled anchors
+        # compute the focal loss
         alpha_factor = keras.backend.ones_like(labels) * alpha
         alpha_factor = keras_retinanet.backend.where(keras.backend.equal(labels, 1), alpha_factor, 1 - alpha_factor)
         focal_weight = keras_retinanet.backend.where(keras.backend.equal(labels, 1), 1 - classification, classification)
         focal_weight = alpha_factor * focal_weight ** gamma
 
         cls_loss = focal_weight * keras.backend.binary_crossentropy(labels, classification)
-        cls_loss = keras.backend.sum(cls_loss)
 
-        # "The total focal loss of an image is computed as the sum
-        # of the focal loss over all ~100k anchors, normalized by the
-        # number of anchors assigned to a ground-truth box."
-        cls_loss = cls_loss / (keras.backend.maximum(1.0, keras.backend.sum(anchor_state)))
-        return cls_loss
+        cls_loss = cls_loss/divisor
 
-    return _focal
+        # filter out "ignore" anchors
+        anchor_state   = keras.backend.max(labels, axis=2)  # -1 for ignore, 0 for background, 1 for object
+        indices        = keras_retinanet.backend.where(keras.backend.not_equal(anchor_state, -1))
+
+        cls_loss = keras_retinanet.backend.gather_nd(cls_loss, indices)
 
+        return keras.backend.sum(cls_loss)/keras.backend.cast(keras.backend.shape(y_true)[0], keras.backend.floatx())
```

Comment:
PEP8

Reply:
Also, isn't this basically `keras.backend.mean` ?

Reply:
No, if you follow the code closely enough you'll see that `cls_loss` doesn't have `batch_size` entries, it has number of entries equal to the number of positive anchors.